Terrastories/terrastories
# Allowing online image and video uploads to S3

This PR includes the configuration necessary to allow the app to use ActiveStorage to upload images and videos (and other media files) to the S3 bucket we have set up. 

Its primary use is for the Heroku demo app, since Heroku does not support a local filesystem.

Generally, we expect development and in-field environments to use the local filesystem for ActiveStorage. You need the encrypted key in order to set the amazon configuration, and I've also included a comment saying not to switch it unless you're sure. So hopefully that dissuades most. :)

I've only included looking up the ENV variable in the production config file anyways, so in theory, you shouldn't be able to use the S3 bucket in development.

Current issue: 
In order for the images to show up properly in the admin dashboard, I needed to install imagemagick on the container itself. Is there a way I can get it to stay there permanently?
I needed to run `docker-compose exec web /bin/bash ` then `apt-get update && apt-get install imagemagick` for it to work.

This branch is currently deployed to Heroku if you'd like to see it in action without pulling!
I didn't have to install imagemagick on Heroku -- it just worked there.

To test these changes:
Pull this branch down.

Ask me to send you the secret keys for S3 config!

Then you'll need to create your Rails credentials (https://www.viget.com/articles/storing-secret-credentials-in-rails-5-2-and-up/)

After that, run these commands to install imagemagick on the container:
```
docker-compose exec web /bin/bash
apt-get update && apt-get install imagemagick
```

If you have media in your database currently, things will break if you change your ENV variable to amazon immediately. You will need to clear out all the rows first!
You can do that by running the following commands:
```
docker-compose exec web /bin/bash 
rails console
Story.delete_all
Speaker.delete_all
Place.delete_all
SpeakerStory.delete_all
```

Next, copy the line in `config/production.rb` and place it in `config/development.rb` in the same spot for active storage.

Change your ENV variable for `ACTIVE_STORAGE_SRC` to `amazon` . Restart the rails container.

You can create some rows in the admin dashboard and upload images and videos! Things should show up in the map and sidebar, and the files should show up in the S3 bucket :sparkles:

Follow ups:
* Rake task that can reset your DB for you (delete all rows)
* Validation for the ENV variable is set to a value that exists in storage.yml
